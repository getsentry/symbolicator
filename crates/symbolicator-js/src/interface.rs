use std::collections::HashSet;
use std::fmt;
use std::sync::Arc;

use reqwest::Url;
use serde::{Deserialize, Serialize};

use symbolicator_service::caching::CacheError;
use symbolicator_service::services::ScrapingConfig;
use symbolicator_service::types::{RawObjectInfo, Scope};
use symbolicator_sources::{SentryFileId, SentrySourceConfig};

#[derive(Debug, Clone)]
pub struct SymbolicateJsStacktraces {
    pub scope: Scope,
    pub source: Arc<SentrySourceConfig>,
    pub release: Option<String>,
    pub dist: Option<String>,
    pub debug_id_index: Option<Url>,
    pub url_index: Option<Url>,
    pub stacktraces: Vec<JsStacktrace>,
    pub modules: Vec<RawObjectInfo>,
    pub scraping: ScrapingConfig,
    /// Whether to apply source context for the stack frames.
    pub apply_source_context: bool,
}

// Some of the renames are there only to make it synchronized
// with the already existing monolith naming scheme.
#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "snake_case")]
#[serde(tag = "type")]
pub enum JsModuleErrorKind {
    InvalidLocation { line: u32, col: Option<u32> },
    InvalidAbsPath,
    NoColumn,
    MissingSourceContent { source: String, sourcemap: String },
    MissingSource,
    MalformedSourcemap { url: String },
    MissingSourcemap,
    InvalidBase64Sourcemap,
    ScrapingDisabled,
}

impl fmt::Display for JsModuleErrorKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            JsModuleErrorKind::InvalidLocation { line, col } => {
                write!(f, "Invalid source location")?;
                match (line, col) {
                    (l, None) => write!(f, ": line:{l}")?,
                    (l, Some(c)) => write!(f, ": line:{l}, col:{c}")?,
                }
                Ok(())
            }
            JsModuleErrorKind::InvalidAbsPath => write!(f, "Invalid absolute path"),
            JsModuleErrorKind::NoColumn => write!(f, "No column information"),
            JsModuleErrorKind::MissingSourceContent { source, sourcemap } => write!(
                f,
                "Missing source contents for source file {source} and sourcemap file {sourcemap}"
            ),
            JsModuleErrorKind::MissingSource => write!(f, "Missing source file"),
            JsModuleErrorKind::MalformedSourcemap { url } => {
                write!(f, "Sourcemap file at {url} is malformed")
            }
            JsModuleErrorKind::MissingSourcemap => write!(f, "Missing sourcemap file"),
            JsModuleErrorKind::InvalidBase64Sourcemap => write!(f, "Invalid base64 sourcemap"),
            JsModuleErrorKind::ScrapingDisabled => {
                write!(f, "Could not download file because scraping is disabled")
            }
        }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq, PartialOrd, Ord)]
pub struct JsModuleError {
    pub abs_path: String,
    #[serde(flatten)]
    pub kind: JsModuleErrorKind,
}

/// An attempt to scrape a JS source or sourcemap file from the web.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct JsScrapingAttempt {
    /// The URL we attempted to scrape from.
    pub url: String,
    /// The outcome of the attempt.
    #[serde(flatten)]
    pub result: JsScrapingResult,
}

impl JsScrapingAttempt {
    pub fn success(url: String) -> Self {
        Self {
            url,
            result: JsScrapingResult::Success,
        }
    }
    pub fn not_attempted(url: String) -> Self {
        Self {
            url,
            result: JsScrapingResult::NotAttempted,
        }
    }

    pub fn failure(url: String, reason: JsScrapingFailureReason, details: String) -> Self {
        Self {
            url,
            result: JsScrapingResult::Failure { reason, details },
        }
    }
}

/// The outcome of a scraping attempt.
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "snake_case")]
#[serde(tag = "status")]
pub enum JsScrapingResult {
    /// We didn't actually attempt scraping because we already obtained the file
    /// by another method.
    NotAttempted,
    /// The file was succesfully scraped.
    Success,
    /// The file couldn't be scraped.
    Failure {
        /// The basic reason for the failure.
        reason: JsScrapingFailureReason,
        #[serde(skip_serializing_if = "String::is_empty")]
        /// A more detailed explanation of the failure.
        details: String,
    },
}

impl From<CacheError> for JsScrapingResult {
    fn from(value: CacheError) -> Self {
        let (reason, details) = match value {
            CacheError::NotFound => (JsScrapingFailureReason::NotFound, String::new()),
            CacheError::PermissionDenied(details) => {
                (JsScrapingFailureReason::PermissionDenied, details)
            }
            CacheError::Timeout(duration) => (
                JsScrapingFailureReason::Timeout,
                format!("Timeout after {}", humantime::format_duration(duration)),
            ),
            CacheError::DownloadError(details) => (JsScrapingFailureReason::DownloadError, details),
            CacheError::Malformed(details) => (JsScrapingFailureReason::Other, details),
            CacheError::InternalError => (JsScrapingFailureReason::Other, String::new()),
        };

        Self::Failure { reason, details }
    }
}

/// The basic reason a scraping attempt failed.
#[derive(Debug, Clone, Copy, Deserialize, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum JsScrapingFailureReason {
    /// The file was not found at the given URL.
    NotFound,
    /// Scraping was disabled.
    Disabled,
    /// The URL was not in the list of allowed hosts or had
    /// an invalid scheme.
    InvalidHost,
    /// Permission to access the file was denied.
    PermissionDenied,
    /// The scraping attempt timed out.
    Timeout,
    /// There was a non-timeout error while downloading.
    DownloadError,
    /// Catchall case.
    ///
    /// This probably can't actually happen.
    Other,
}

#[derive(Debug, Default, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct JsFrame {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub function: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub filename: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub module: Option<String>,

    pub abs_path: String,

    pub lineno: u32,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub colno: Option<u32>,

    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub pre_context: Vec<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub context_line: Option<String>,

    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub post_context: Vec<String>,

    #[serde(skip_serializing)]
    pub token_name: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub in_app: Option<bool>,

    #[serde(default, skip_serializing_if = "JsFrameData::is_empty")]
    pub data: JsFrameData,
}

#[derive(Debug, Default, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct JsFrameData {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sourcemap: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resolved_with: Option<ResolvedWith>,
    #[serde(default)]
    pub symbolicated: bool,
}

/// A marker indicating what a File was resolved with.
///
/// This enum serves a double purpose, both marking how an individual file was found inside of a
/// bundle, as well as tracking through which method that bundle itself was found.
///
#[derive(Debug, Default, Clone, Copy, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub enum ResolvedWith {
    /// Both: Found in a Bundle via DebugId
    /// And: Found the Bundle via API Lookup via DebugId / Database Index
    DebugId,
    /// Found in a Bundle via Url matching
    Url,
    /// Found the Bundle via API Lookup via Database Index
    Index,
    /// Found the File in a Flat File / Bundle Index
    BundleIndex,
    /// Found the Bundle via API Lookup as an ArtifactBundle
    Release,
    /// Found the Bundle via API Lookup as a ReleaseFile
    ReleaseOld,
    /// Scraped the File from the Web
    Scraping,
    /// Unknown
    #[default]
    Unknown,
}

impl JsFrameData {
    pub fn is_empty(&self) -> bool {
        *self == Self::default()
    }
}

#[derive(Clone, Debug, Default, Deserialize, Serialize)]
pub struct JsStacktrace {
    pub frames: Vec<JsFrame>,
}

#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct CompletedJsSymbolicationResponse {
    pub stacktraces: Vec<JsStacktrace>,
    pub raw_stacktraces: Vec<JsStacktrace>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub errors: Vec<JsModuleError>,
    #[serde(skip_serializing_if = "HashSet::is_empty")]
    pub used_artifact_bundles: HashSet<SentryFileId>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub scraping_attempts: Vec<JsScrapingAttempt>,
}
