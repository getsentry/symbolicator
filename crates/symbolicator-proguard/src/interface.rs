use std::sync::Arc;

use serde::{Deserialize, Serialize};
use symbolic::common::DebugId;
use symbolicator_service::types::Scope;
use symbolicator_sources::SourceConfig;

/// A request for symbolication/remapping of a JVM event.
#[derive(Debug, Clone)]
pub struct SymbolicateJvmStacktraces {
    /// The scope of this request which determines access to cached files.
    pub scope: Scope,
    /// A list of external sources to load debug files.
    pub sources: Arc<[SourceConfig]>,
    /// The exceptions to symbolicate/remap.
    pub exceptions: Vec<JvmException>,
    /// The list of stacktraces to symbolicate/remap.
    pub stacktraces: Vec<JvmStacktrace>,
    /// A list of proguard files to use for remapping.
    pub modules: Vec<JvmModule>,
    /// Whether to apply source context for the stack frames.
    pub apply_source_context: bool,
    /// The package name of the release this event belongs to.
    ///
    /// This is used to set a remapped frame's `in_app` field.
    pub release_package: Option<String>,
}

/// A stack frame in a JVM stacktrace.
#[derive(Debug, Default, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct JvmFrame {
    /// The frame's function name.
    ///
    /// For a JVM frame, this is always a class method.
    pub function: String,

    /// The source file name.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub filename: Option<String>,

    /// The frame's method name.
    ///
    /// For a JVM frame, this is a fully qualified class name.
    pub module: String,

    /// The source file's absolute path.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub abs_path: Option<String>,

    /// The line number within the source file, starting at `1` for the first line.
    pub lineno: u32,

    /// Source context before the context line.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub pre_context: Vec<String>,

    /// The context line if available.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context_line: Option<String>,

    /// Post context after the context line
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub post_context: Vec<String>,

    /// Whether the frame is related to app-code (rather than libraries/dependencies).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub in_app: Option<bool>,
}

/// An exception in a JVM event.
#[derive(Debug, Default, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct JvmException {
    /// The type (class name) of the exception.
    #[serde(rename = "type")]
    pub ty: String,
    /// The module in which the exception is defined.
    pub module: String,
}

/// A JVM stacktrace.
#[derive(Debug, Default, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct JvmStacktrace {
    /// The stacktrace's frames.
    pub frames: Vec<JvmFrame>,
}

/// A JVM module (proguard file).
#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct JvmModule {
    /// The file's UUID.
    ///
    /// This is used to download the file from symbol sources.
    pub uuid: DebugId,
}

// TODO: Expand this
/// The symbolicated/remapped event data.
#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct CompletedJvmSymbolicationResponse {
    /// The exceptions after remapping.
    pub exceptions: Vec<JvmException>,
    /// The stacktraces after remapping.
    pub stacktraces: Vec<JvmStacktrace>,
    /// Errors that occurred during symbolication.
    pub errors: Vec<(DebugId, String)>,
}
