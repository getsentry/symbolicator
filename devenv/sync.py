from devenv import constants
from devenv.lib import config, proc, uv


def main(context: dict[str, str]) -> int:
    reporoot = context["reporoot"]
    cfg = config.get_repo(reporoot)

    uv.install(
        cfg["uv"]["version"],
        cfg["uv"][constants.SYSTEM_MACHINE],
        cfg["uv"][f"{constants.SYSTEM_MACHINE}_sha256"],
        reporoot,
    )

    print("syncing .venv ...")
    proc.run(
        (f"{reporoot}/.devenv/bin/uv", "sync", "--frozen", "--quiet", "--active"),
    )

    print("installing pre-commit hooks ...")
    proc.run((f"{reporoot}/.venv/bin/pre-commit", "install", "--install-hooks"))

    return 0
