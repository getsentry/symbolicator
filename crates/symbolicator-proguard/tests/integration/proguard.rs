use std::sync::Arc;
use std::{collections::HashMap, str::FromStr};

use serde_json::json;
use symbolic::common::{DebugId, Uuid};
use symbolicator_proguard::interface::{
    CompletedJvmSymbolicationResponse, JvmException, JvmFrame, JvmModule, JvmModuleType,
    JvmStacktrace, SymbolicateJvmStacktraces,
};
use symbolicator_service::types::Scope;
use symbolicator_sources::{SentrySourceConfig, SourceConfig};

use crate::setup_service;

fn proguard_server<L>(
    fixtures_dir: &str,
    lookup: L,
) -> (symbolicator_test::Server, SentrySourceConfig)
where
    L: Fn(&str, &HashMap<String, String>) -> serde_json::Value + Clone + Send + 'static,
{
    let fixtures_dir = symbolicator_test::fixture(format!("proguard/{fixtures_dir}"));
    symbolicator_test::sentry_server(fixtures_dir, lookup)
}

#[tokio::test]
async fn test_download_proguard_file() {
    symbolicator_test::setup();
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) = proguard_server("download_proguard_file", |_url, _query| {
        json!([{
            "id":"proguard.txt",
            "uuid":"246fb328-fc4e-406a-87ff-fc35f6149d8f",
            "debugId":"246fb328-fc4e-406a-87ff-fc35f6149d8f",
            "codeId":null,
            "cpuName":"any",
            "objectName":"proguard-mapping",
            "symbolType":"proguard",
            "headers": {
                "Content-Type":"text/x-proguard+plain"
            },
            "size":3619,
            "sha1":"deba83e73fd18210a830db372a0e0a2f2293a989",
            "dateCreated":"2024-02-14T10:49:38.770116Z",
            "data":{
                "features":["mapping"]
            }
        }])
    });

    let source = SourceConfig::Sentry(Arc::new(source));
    let debug_id = DebugId::from_str("246fb328-fc4e-406a-87ff-fc35f6149d8f").unwrap();

    assert!(symbolication
        .download_proguard_file(&[source], &Scope::Global, debug_id)
        .await
        .is_ok());
}

#[tokio::test]
async fn test_remap_exception() {
    symbolicator_test::setup();
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) = proguard_server("remap_exception", |_url, _query| {
        json!([{
            "id":"proguard.txt",
            "uuid":"246fb328-fc4e-406a-87ff-fc35f6149d8f",
            "debugId":"246fb328-fc4e-406a-87ff-fc35f6149d8f",
            "codeId":null,
            "cpuName":"any",
            "objectName":"proguard-mapping",
            "symbolType":"proguard",
            "headers": {
                "Content-Type":"text/x-proguard+plain"
            },
            "size":3619,
            "sha1":"deba83e73fd18210a830db372a0e0a2f2293a989",
            "dateCreated":"2024-02-14T10:49:38.770116Z",
            "data":{
                "features":["mapping"]
            }
        }])
    });

    let source = SourceConfig::Sentry(Arc::new(source));
    let debug_id = DebugId::from_str("246fb328-fc4e-406a-87ff-fc35f6149d8f").unwrap();

    let exception = JvmException {
        ty: "g$a".into(),
        module: "org.a.b".into(),
    };

    let request = SymbolicateJvmStacktraces {
        scope: Scope::Global,
        sources: Arc::new([source]),
        exceptions: vec![exception.clone()],
        stacktraces: vec![],
        modules: vec![JvmModule {
            uuid: debug_id,
            r#type: JvmModuleType::Proguard,
        }],
        apply_source_context: false,
        release_package: None,
    };

    let response = symbolication.symbolicate_jvm(request).await;

    assert_snapshot!(response);
}

#[tokio::test]
async fn test_resolving_inline() {
    symbolicator_test::setup();
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) = proguard_server("resolving_inline", |_url, _query| {
        json!([{
            "id":"proguard.txt",
            "uuid":"246fb328-fc4e-406a-87ff-fc35f6149d8f",
            "debugId":"246fb328-fc4e-406a-87ff-fc35f6149d8f",
            "codeId":null,
            "cpuName":"any",
            "objectName":"proguard-mapping",
            "symbolType":"proguard",
            "headers": {
                "Content-Type":"text/x-proguard+plain"
            },
            "size":3619,
            "sha1":"deba83e73fd18210a830db372a0e0a2f2293a989",
            "dateCreated":"2024-02-14T10:49:38.770116Z",
            "data":{
                "features":["mapping"]
            }
        }])
    });

    let source = SourceConfig::Sentry(Arc::new(source));
    let debug_id = DebugId::from_str("246fb328-fc4e-406a-87ff-fc35f6149d8f").unwrap();

    let exception = JvmException {
        ty: "g$a".into(),
        module: "org.a.b".into(),
    };

    let stacktraces = vec![JvmStacktrace {
        frames: vec![
            JvmFrame {
                function: "onClick".into(),
                module: "e.a.c.a".into(),
                lineno: 2,
                index: 0,
                ..Default::default()
            },
            JvmFrame {
                function: "t".into(),
                module: "io.sentry.sample.MainActivity".into(),
                filename: Some("MainActivity.java".into()),
                lineno: 1,
                index: 1,
                ..Default::default()
            },
        ],
    }];

    let request = SymbolicateJvmStacktraces {
        scope: Scope::Global,
        sources: Arc::new([source]),
        exceptions: vec![exception.clone()],
        stacktraces,
        modules: vec![JvmModule {
            uuid: debug_id,
            r#type: JvmModuleType::Proguard,
        }],
        apply_source_context: false,
        release_package: None,
    };

    let response = symbolication.symbolicate_jvm(request).await;

    assert_snapshot!(response);
}

#[tokio::test]
async fn test_basic_source_lookup() {
    symbolicator_test::setup();
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) = proguard_server("basic_source_lookup", |_url, _query| {
        json!([{
            "id":"bundle.zip",
            "uuid":"246fb328-fc4e-406a-87ff-fc35f6149d8f",
            "debugId":"246fb328-fc4e-406a-87ff-fc35f6149d8f",
            "codeId":null,
            "cpuName":"any",
            "objectName":"proguard-mapping",
            "symbolType":"sourcebundle",
            "headers": {
                "Content-Type":"application/x-sentry-bundle+zip"
            },
            "size":3619,
            "sha1":"deba83e73fd18210a830db372a0e0a2f2293a989",
            "dateCreated":"2024-02-14T10:49:38.770116Z",
        }])
    });

    let source = SourceConfig::Sentry(Arc::new(source));
    let debug_id = DebugId::from_str("246fb328-fc4e-406a-87ff-fc35f6149d8f").unwrap();

    let exception = JvmException {
        ty: "RuntimeException".into(),
        module: "io.sentry.samples".into(),
    };

    let stacktraces = vec![JvmStacktrace {
        frames: vec![
            JvmFrame {
                function: "otherMethod".into(),
                abs_path: Some("OtherActivity.java".into()),
                module: "OtherActivity".into(),
                filename: Some("OtherActivity.java".into()),
                lineno: 100,
                index: 0,
                ..Default::default()
            },
            JvmFrame {
                function: "differentMethod".into(),
                abs_path: Some("DifferentActivity".into()),
                module: "DifferentActivity".into(),
                filename: Some("DifferentActivity".into()),
                lineno: 200,
                index: 1,
                ..Default::default()
            },
            JvmFrame {
                function: "onCreate".into(),
                abs_path: None,
                module: "io.sentry.samples.MainActivity".into(),
                filename: None,
                lineno: 11,
                index: 2,
                ..Default::default()
            },
            JvmFrame {
                function: "whoops".into(),
                abs_path: Some("MainActivity.kt".into()),
                module: "io.sentry.samples.MainActivity$InnerClass".into(),
                filename: Some("MainActivity.kt".into()),
                lineno: 20,
                index: 3,
                ..Default::default()
            },
            JvmFrame {
                function: "whoops2".into(),
                abs_path: None,
                module: "io.sentry.samples.MainActivity$AnotherInnerClass".into(),
                filename: None,
                lineno: 26,
                index: 4,
                ..Default::default()
            },
            JvmFrame {
                function: "whoops3".into(),
                abs_path: Some("MainActivity.kt".into()),
                module: "io.sentry.samples.MainActivity$AdditionalInnerClass".into(),
                filename: Some("MainActivity.kt".into()),
                lineno: 32,
                index: 5,
                ..Default::default()
            },
            JvmFrame {
                function: "whoops4".into(),
                abs_path: Some("SourceFile".into()),
                module: "io.sentry.samples.MainActivity$OneMoreInnerClass".into(),
                filename: Some("SourceFile".into()),
                lineno: 38,
                index: 6,
                ..Default::default()
            },
        ],
    }];

    let request = SymbolicateJvmStacktraces {
        scope: Scope::Global,
        sources: Arc::new([source]),
        exceptions: vec![exception.clone()],
        stacktraces,
        modules: vec![JvmModule {
            uuid: debug_id,
            r#type: JvmModuleType::Source,
        }],
        apply_source_context: true,
        release_package: None,
    };

    let response = symbolication.symbolicate_jvm(request).await;

    assert_snapshot!(response);
}

#[tokio::test]
async fn test_source_lookup_with_proguard() {
    symbolicator_test::setup();
    let (symbolication, _cache_dir) = setup_service(|_| ());

    let proguard_id = "05d96b1c-1786-477c-8615-d3cf83e027c7".parse().unwrap();
    let missing_proguard_id = "8236f5cf-52c8-4e35-a7cf-01421e4c2c88".parse().unwrap();
    let source_id1 = DebugId::from_uuid(Uuid::new_v4());
    let source_id2 = DebugId::from_uuid(Uuid::new_v4());

    let (_srv, source) = proguard_server("source_lookup_with_proguard", move |_url, query| {
        let debug_id: DebugId = query["debug_id"].parse().unwrap();

        match debug_id {
            _ if debug_id == source_id1 => {
                json!([{
                    "id":"edit_activity.zip",
                    "debugId":source_id1,
                    "codeId":null,
                    "cpuName":"any",
                    "objectName":"proguard-mapping",
                    "symbolType":"sourcebundle",
                    "headers": {
                        "Content-Type":"application/x-sentry-bundle+zip"
                    },
                    "size":3619,
                    "sha1":"deba83e73fd18210a830db372a0e0a2f2293a989",
                    "dateCreated":"2024-02-14T10:49:38.770116Z",
                }])
            }
            _ if debug_id == source_id2 => {
                json!([{
                    "id":"some_service.zip",
                    "debugId":source_id2,
                    "codeId":null,
                    "cpuName":"any",
                    "objectName":"proguard-mapping",
                    "symbolType":"sourcebundle",
                    "headers": {
                        "Content-Type":"application/x-sentry-bundle+zip"
                    },
                    "size":3619,
                    "sha1":"deba83e73fd18210a830db372a0e0a2f2293a989",
                    "dateCreated":"2024-02-14T10:49:38.770116Z",
                }])
            }
            _ if debug_id == proguard_id => {
                json!([{
                    "id":"proguard.txt",
                    "uuid":proguard_id,
                    "debugId":proguard_id,
                    "codeId":null,
                    "cpuName":"any",
                    "objectName":"proguard-mapping",
                    "symbolType":"proguard",
                    "headers": {
                        "Content-Type":"text/x-proguard+plain"
                    },
                    "size":3619,
                    "sha1":"deba83e73fd18210a830db372a0e0a2f2293a989",
                    "dateCreated":"2024-02-14T10:49:38.770116Z",
                    "data":{
                        "features":["mapping"]
                    }
                }])
            }
            _ => json!([]),
        }
    });

    let source = SourceConfig::Sentry(Arc::new(source));

    let exception = JvmException {
        ty: "RuntimeException".into(),
        module: "java.lang".into(),
    };

    let stacktraces = vec![JvmStacktrace {
        frames: vec![
            JvmFrame {
                filename: Some("ZygoteInit.java".into()),
                function: "main".into(),
                module: "com.android.internal.os.ZygoteInit".into(),
                lineno: 698,
                index: 0,
                ..Default::default()
            },
            JvmFrame {
                filename: Some("ZygoteInit.java".into()),
                function: "run".into(),
                module: "com.android.internal.os.ZygoteInit$MethodAndArgsCaller".into(),
                lineno: 903,
                index: 1,
                ..Default::default()
            },
            JvmFrame {
                filename: Some("Method.java".into()),
                function: "invoke".into(),
                module: "java.lang.reflect.Method".into(),
                lineno: 372,
                index: 2,
                ..Default::default()
            },
            JvmFrame {
                filename: Some("Method.java".into()),
                function: "invoke".into(),
                module: "java.lang.reflect.Method".into(),
                index: 3,
                ..Default::default()
            },
            JvmFrame {
                filename: Some("ActivityThread.java".into()),
                function: "main".into(),
                module: "android.app.ActivityThread".into(),
                lineno: 5254,
                index: 4,
                ..Default::default()
            },
            JvmFrame {
                filename: Some("Looper.java".into()),
                function: "loop".into(),
                module: "android.os.Looper".into(),
                lineno: 135,
                index: 5,
                ..Default::default()
            },
            JvmFrame {
                filename: Some("Handler.java".into()),
                function: "dispatchMessage".into(),
                module: "android.os.Handler".into(),
                lineno: 95,
                index: 6,
                ..Default::default()
            },
            JvmFrame {
                filename: Some("Handler.java".into()),
                function: "handleCallback".into(),
                module: "android.os.Handler".into(),
                lineno: 739,
                index: 7,
                ..Default::default()
            },
            JvmFrame {
                filename: Some("View.java".into()),
                function: "run".into(),
                module: "android.view.View$PerformClick".into(),
                lineno: 19866,
                index: 8,
                ..Default::default()
            },
            JvmFrame {
                filename: Some("View.java".into()),
                function: "performClick".into(),
                module: "android.view.View".into(),
                lineno: 4780,
                index: 9,
                ..Default::default()
            },
            JvmFrame {
                filename: Some("ActionMenuItemView.java".into()),
                function: "onClick".into(),
                module: "androidx.appcompat.view.menu.ActionMenuItemView".into(),
                lineno: 7,
                index: 10,
                ..Default::default()
            },
            JvmFrame {
                filename: Some("ActionMenuView.java".into()),
                function: "invokeItem".into(),
                module: "androidx.appcompat.widget.ActionMenuView".into(),
                lineno: 4,
                index: 11,
                ..Default::default()
            },
            JvmFrame {
                filename: Some("MenuBuilder.java".into()),
                function: "performItemAction".into(),
                module: "androidx.appcompat.view.menu.MenuBuilder".into(),
                lineno: 1,
                index: 12,
                ..Default::default()
            },
            JvmFrame {
                filename: Some("MenuBuilder.java".into()),
                function: "performItemAction".into(),
                module: "androidx.appcompat.view.menu.MenuBuilder".into(),
                lineno: 4,
                index: 13,
                ..Default::default()
            },
            JvmFrame {
                filename: Some("MenuItemImpl.java".into()),
                function: "invoke".into(),
                module: "androidx.appcompat.view.menu.MenuItemImpl".into(),
                lineno: 15,
                index: 14,
                ..Default::default()
            },
            JvmFrame {
                filename: Some("MenuBuilder.java".into()),
                function: "dispatchMenuItemSelected".into(),
                module: "androidx.appcompat.view.menu.MenuBuilder".into(),
                lineno: 5,
                index: 15,
                ..Default::default()
            },
            JvmFrame {
                filename: Some("ActionMenuView.java".into()),
                function: "onMenuItemSelected".into(),
                module: "androidx.appcompat.widget.ActionMenuView$MenuBuilderCallback".into(),
                lineno: 7,
                index: 16,
                ..Default::default()
            },
            JvmFrame {
                filename: Some("Toolbar.java".into()),
                function: "onMenuItemClick".into(),
                module: "androidx.appcompat.widget.Toolbar$1".into(),
                lineno: 7,
                index: 17,
                ..Default::default()
            },
            JvmFrame {
                filename: Some("R8$$SyntheticClass".into()),
                function: "onMenuItemClick".into(),
                module: "io.sentry.samples.instrumentation.ui.g".into(),
                lineno: 40,
                in_app: Some(true),
                index: 18,
                ..Default::default()
            },
        ],
    }];

    let modules = vec![
        JvmModule {
            r#type: JvmModuleType::Source,
            uuid: source_id1,
        },
        JvmModule {
            r#type: JvmModuleType::Source,
            uuid: source_id2,
        },
        JvmModule {
            r#type: JvmModuleType::Proguard,
            uuid: proguard_id,
        },
        JvmModule {
            r#type: JvmModuleType::Proguard,
            uuid: missing_proguard_id,
        },
    ];

    let request = SymbolicateJvmStacktraces {
        scope: Scope::Global,
        sources: Arc::new([source]),
        exceptions: vec![exception.clone()],
        stacktraces,
        modules,
        apply_source_context: true,
        release_package: None,
    };

    let response = symbolication.symbolicate_jvm(request).await;

    assert_snapshot!(response);
}
