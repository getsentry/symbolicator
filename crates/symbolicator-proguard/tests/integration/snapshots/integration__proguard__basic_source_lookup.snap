---
source: crates/symbolicator-proguard/tests/integration/proguard.rs
assertion_line: 517
expression: response
---
exceptions:
  - type: RuntimeException
    module: io.sentry.samples
stacktraces:
  - frames:
      - function: otherMethod
        filename: OtherActivity.java
        module: OtherActivity
        abs_path: OtherActivity.java
        lineno: 100
        index: 0
      - function: differentMethod
        filename: DifferentActivity
        module: DifferentActivity
        abs_path: DifferentActivity
        lineno: 200
        index: 1
      - function: onCreate
        module: io.sentry.samples.MainActivity
        lineno: 11
        pre_context:
          - ""
          - "class MainActivity : ComponentActivity() {"
          - "    override fun onCreate(savedInstanceState: Bundle?) {"
          - "        super.onCreate(savedInstanceState)"
          - "        setContentView(R.layout.activity_main)"
        context_line: "        InnerClass().whoops()"
        post_context:
          - ""
          - "        val list = findViewById<RecyclerView>(R.id.list)"
          - "        list.layoutManager = LinearLayoutManager(this)"
          - "        list.adapter = TrackAdapter()"
          - "    }"
        index: 2
      - function: whoops
        filename: MainActivity.kt
        module: io.sentry.samples.MainActivity$InnerClass
        abs_path: MainActivity.kt
        lineno: 20
        pre_context:
          - "        list.adapter = TrackAdapter()"
          - "    }"
          - ""
          - "    class InnerClass {"
          - "        fun whoops() {"
        context_line: "            AnotherInnerClass().whoops2()"
        post_context:
          - "        }"
          - "    }"
          - ""
          - "    class AnotherInnerClass {"
          - "        fun whoops2() {"
        index: 3
      - function: whoops2
        module: io.sentry.samples.MainActivity$AnotherInnerClass
        lineno: 26
        pre_context:
          - "        }"
          - "    }"
          - ""
          - "    class AnotherInnerClass {"
          - "        fun whoops2() {"
        context_line: "            AdditionalInnerClass().whoops3()"
        post_context:
          - "        }"
          - "    }"
          - ""
          - "    class AdditionalInnerClass {"
          - "        fun whoops3() {"
        index: 4
      - function: whoops3
        filename: MainActivity.kt
        module: io.sentry.samples.MainActivity$AdditionalInnerClass
        abs_path: MainActivity.kt
        lineno: 32
        pre_context:
          - "        }"
          - "    }"
          - ""
          - "    class AdditionalInnerClass {"
          - "        fun whoops3() {"
        context_line: "            OneMoreInnerClass().whoops4()"
        post_context:
          - "        }"
          - "    }"
          - ""
          - "    class OneMoreInnerClass {"
          - "        fun whoops4() {"
        index: 5
      - function: whoops4
        filename: SourceFile
        module: io.sentry.samples.MainActivity$OneMoreInnerClass
        abs_path: SourceFile
        lineno: 38
        pre_context:
          - "        }"
          - "    }"
          - ""
          - "    class OneMoreInnerClass {"
          - "        fun whoops4() {"
        context_line: "            throw RuntimeException(\"whoops\")"
        post_context:
          - "        }"
          - "    }"
          - "}"
          - ""
        index: 6
errors: []
